README.md
README
================
Yue Pan
19 April, 2022

## Overview

The R package ‘scpoisson’ is developed to visualize the Poissoneity of
scRNA-seq data, and explore cell clustering based on model departure as
a novel data representation.

## Installation

Please install from bitbucket repository “dittmerlab/scpoisson”.

``` r
remotes::install_bitbucket("dittmerlab/scpoisson")
```

``` r
library(scpoisson)
library(magrittr)
library(purrr)
```

    ## 
    ## Attaching package: 'purrr'

    ## The following object is masked from 'package:magrittr':
    ## 
    ##     set_names

## Visualization

The examples below show Q-Q envelope plots comparing sample data with a
theoretical Poisson distribution given the Poisson parameter (mean). The
input can be either a numeric vector or a numeric matrix consists of
integers. If the input is a numeric matrix, the GLM-PCA algorithm will
be applied for parameter estimation. Then some matrix entries (default
200, with estimated Poisson parameters closest to the given Poisson
parameter) will be selected as sample data and compare with the
theoretical distribution. If the theoretical distribution fits the
sample data well, the quantile points (or line) will approximately lie
on the diagonal line, and also within the envelope.

The code here is specific for validation of independent Poisson
distributions, but such idea can be applied to different types of data
under different assumptions of distributions.

``` r
# Numeric vector as input
set.seed(1234)
scppp_obj <- scppp(rpois(200, 3))
qqplot_env_pois(scppp_obj, 3, 100)
```

![](README_files/figure-gfm/example1-1.png)<!-- -->

``` r
# Count matrix as input
set.seed(1234)
dat <- matrix(c(rpois(300, 5), rpois(200, 1)), ncol = 20)
scppp_obj <- scppp(dat)
qqplot_env_pois(scppp_obj, L = 2, lambda = 5) # small L for a data with less variance
```

![](README_files/figure-gfm/example1-2.png)<!-- -->

Please refer to vignette for a more detailed guide, including the
application in the real data.

## Clustering

The example below shows Hclust-Depart clustering pipeline based on based
on model departure using a simplified two-cluster simulated data. The
results will be saved in a list. The first element contains the
clustering results for each sample, where the cluster label shows the
top-down cluster in a tree structure. E.g. Both cluster ‘1-1’ and
cluster ‘1-2’ are subclusters from cluster ‘1’ after first split; while
cluster ‘1-1’ and cluster ‘2-1’ come from different cluster after first
split. The second element is a matrix with samples/cells as rows and
cluster index as columns, where each entry contains the hypothesis
testing p-value corresponding to each cluster at each split step (each
cell within the that cluster will have the same p-value).

This simplified example only contains two clusters, with cluster label
either ‘1’ or ‘2’.

``` r
set.seed(1234)
para1 <- matrix(c(rep(2L, 20), rep(1L, 40)), ncol = 1)
para2 <- matrix(c(rep(2L, 10), rep(1L, 20)), nrow = 1)
dat <- para1 %*% para2
dat[31:60, 1:10] <- dat[31:60, 1:10] + 10L
counts <- map_int(dat, ~rpois(1, .x)) %>% 
    matrix(ncol = 30)
colnames(counts) <- paste0("cell", 1:30)
rownames(counts) <- paste0("gene", 1:60)
scppp_obj <- scppp(counts)
heatmap(scppp_obj[["data"]])
```

![](README_files/figure-gfm/example2-1.png)<!-- -->

``` r
scppp_obj <- HclustDepart(scppp_obj, maxSplit = 3)

# cluster results for each cell after each split
head(scppp_obj[["clust_results"]]$Hclust[[1]])
```

    ##   names cluster
    ## 1 cell1       1
    ## 2 cell2       1
    ## 3 cell3       1
    ## 4 cell4       1
    ## 5 cell5       1
    ## 6 cell6       1

``` r
# p-value for each cell after each split
head(scppp_obj[["clust_results"]]$Hclust[[2]])
```

    ##         [,1] [,2] [,3]
    ## [1,] 0.02274   NA   NA
    ## [2,] 0.02274   NA   NA
    ## [3,] 0.02274   NA   NA
    ## [4,] 0.02274   NA   NA
    ## [5,] 0.02274   NA   NA
    ## [6,] 0.02274   NA   NA

Please refer to vignette for a more detailed guide, including the
application in the real data.

Another option is: keep model departure as data representation but apply
the Louvain algorithm (implemented in Seurat pipeline) for clustering.

``` r
scppp_obj <- LouvainDepart(scppp_obj)
```

    ## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
    ## 
    ## Number of nodes: 30
    ## Number of edges: 435
    ## 
    ## Running Louvain algorithm...
    ## Maximum modularity in 10 random starts: 0.2517
    ## Number of communities: 2
    ## Elapsed time: 0 seconds

``` r
head(scppp_obj[["clust_results"]]$Lclust[[4]])
```

    ##       names cluster
    ## cell1 cell1       1
    ## cell2 cell2       1
    ## cell3 cell3       1
    ## cell4 cell4       1
    ## cell5 cell5       1
    ## cell6 cell6       1

In this simple case, both clustering pipeline give the same clustering
results as we expected.

## Differential Expression

We can find differentially expressed genes between two clusters based on
model departure. Run ‘adj\_CDF\_logit’ function first to calculate the
departure representation. The cluster labels for such comparison need to
match the output from Hclust-Depart clustering. The genes in the output
will be ranked by decreasing order of mean difference between two
clusters.

``` r
scppp_obj <- adj_CDF_logit(scppp_obj) 
scppp_obj <- diff_gene_list(scppp_obj, clust1 = "1", clust2 = "2", t = F)
head(scppp_obj[["de_results"]]$Hclust)
```

    ##   variable clust1_mean clust2_mean clust1_n clust2_n mean_diff statistic
    ## 1   gene42    1.145131   -1.346082       10       20  2.491212       177
    ## 2   gene39    1.107470   -1.314327       10       20  2.421796       183
    ## 3   gene48    1.098889   -1.291513       10       20  2.390401       184
    ## 4   gene53    1.048909   -1.245737       10       20  2.294646       175
    ## 5   gene58    1.044325   -1.250291       10       20  2.294616       166
    ## 6   gene52    1.042028   -1.244017       10       20  2.286046       181
    ##        p.value        padj abs_diff
    ## 1 0.0007587375 0.002845266 2.491212
    ## 2 0.0002830659 0.001887106 2.421796
    ## 3 0.0002369525 0.001777144 2.390401
    ## 4 0.0010403270 0.002851969 2.294646
    ## 5 0.0039363264 0.007618696 2.294616
    ## 6 0.0003972232 0.002018305 2.286046

## Issues and bug reports

Please report any issues at [bitbucket
repository](https://bitbucket.org/dittmerlab/scpoisson/issues).

## References

Townes, F. W., Hicks, S. C., Aryee, M. J., & Irizarry, R. A. (2019).
Feature selection and dimension reduction for single-cell RNA-Seq based
on a multinomial model. Genome biology, 20(1), 1-16.

Stuart T, Butler A, Hoffman P, Hafemeister C, Papalexi E, III WMM, Hao
Y, Stoeckius M, Smibert P, Satija R (2019). “Comprehensive Integration
of Single-Cell Data.” Cell, 177, 1888-1902. doi:
10.1016/j.cell.2019.05.031,
<https://doi.org/10.1016/j.cell.2019.05.031>.


NEWS.md

# scpoisson 0.0.1

Initial CRAN release


R/qq_env_small.R
#' A more "continuous" approximation of quantiles of samples with a few integer case
#'
#' This function returns a data frame including data points and corresponding quantile.
#'
#' This is a function developed to get quantile for samples with only a few integer values.
#' Define both \eqn{p_{-1} = 0} and \eqn{q_{-1} = 0}.
#' Replace the point mass at each integer \eqn{z} by a bar on the interval \eqn{[z – \frac{1}{2}, z+ \frac{1}{2}]}
#' with height \eqn{P(X = z)}. This is a more "continuous" approximation of quantiles in this case.
#'
#' @param data A numeric vector of sampled data points.
#' @param sample A character string denotes which sample data points come from.
#'
#' @return A data frame contains the corresponding probability from cumulative distribution function (CDF), sample name, and corresponding respective quantiles.
#'
#'
new_quantile <- function(data, sample){

  if(!is.numeric(data)){
    warning("need numeric values for input data")
  }

  dfp <- data.frame(value = data)
  dfp <- dfp %>%
    dplyr::do(data.frame(.data, fval = ecdf(.data$value)(.data$value)))
  dfp <- dfp %>% distinct()
  dfp <- dfp[with(dfp, order(fval)),]
  dfp$cat <- sample
  dfp <- rbind(data.frame(dfp), data.frame(value = c(min(data)-1), cat = sample, fval = c(0)))
  dfp$value_new <- dfp$value + 1/2
  dfp <- dfp[with(dfp, order(fval)),]
  return(dfp)
}

#' Linear interpolation for one sample given reference sample
#'
#' This function returns a data frame with interpolated data points.
#'
#' This is a function developed to do linear interpolation for corresponding probability
#' from empirical cumulative distribution function (CDF) and corresponding quantiles.
#' Given a reference data frame and a data frame needed to do interpolation,
#' if there are any CDF values in reference but not in object data frame,
#' do the linear interpolation and insert both CDF values and respective quantiles
#' to the original object data frame.
#'
#' @param df The object data frame requires interpolation.
#' @param reference The reference data frame to make comparison.
#' @param sample_id A character to denote the object data frame.
#'
#' @return A data frame contains CDF, the sample name, and the corresponding quantiles.
#'
#'
interpolate <- function(df, reference, sample_id){
  list_ <- c(list(df), lapply(1:nrow(reference), function(x) x))
  df <- purrr::reduce(list_, function(df, i, reference){
    if (df$fval[1] == 0 &reference$fval[i] < 1 & reference$fval[i] %notin% df$fval){
      cat_add <- sample_id
      fval_add <- reference$fval[i]
      index <- max(which(df$fval < fval_add))
      if (index < nrow(df))
      {value_new_add <- approx(df$fval[index : (index+1)], df$value_new[index : (index+1)], fval_add)$y
      df <- rbind(data.frame(df), data.frame(value = (value_new_add-1/2), fval = fval_add, cat = cat_add, value_new = value_new_add))}
      df <- df[with(df, order(fval)),]
    }
    return(df)
  }, reference = reference)
  return(df)
}

#' Paired quantile after interpolation between two samples
#'
#' This function returns a data frame with paired quantiles in two samples after interpolation.
#'
#' This is a function for quantile interpolation of two samples.
#' For each unique quantile value that has original data
#' point in one sample but no corresponding original data point in another sample,
#' apply a linear interpolation. So the common quantile values after interpolation
#' should have unique points the same as unique quantile points from either sample.
#'
#' @param dfp A data frame generated from function new_quantile() based on a specific distribution.
#' @param dfq Another data frame generated from function new_quantile() based on a specific distribution.
#' @param sample1 A character to denote sample name of distribution used to generate \code{dfp}.
#' @param sample2 A character to denote sample name of distribution used to generate \code{dfq}.
#'
#' @return A data frame contains corresponding probability from cumulative distribution function (CDF),
#' corresponding quantiles from the first sample (\code{dfp}),
#' and corresponding quantiles from the second sample (\code{dfq}).
#'
#'
qq_interpolation <- function(dfp, dfq, sample1, sample2){


  if (!all(colnames(dfp) == c("value", "fval", "cat", "value_new"))){
    warning("column names for first data not match, run new_quantile() function first")
  }

  if (!all(colnames(dfq) == c("value", "fval", "cat", "value_new"))){
    warning("column names for second data not match, run new_quantile() function first")
  }

  if (dfp$fval[1] != 0){
    warning("fvalue for first data not start from 0, run new_quantile() function first")
  }

  if (dfq$fval[1] != 0){
    warning("fvalue for second data not start from 0, run new_quantile() function first")
  }

  # interpolation on dfq
  dfq <- interpolate(df = dfq, reference = dfp, sample_id = sample2)

  # interpolation on dfp
  dfp <- interpolate(df = dfp, reference = dfq, sample_id = sample1)

  colnames(dfp) <- c("value", "fval", "cat", "value_new_p")
  dfp1 <- dfp[,c("fval", "value_new_p")]
  colnames(dfq) <- c("value", "fval", "cat", "value_new_q")
  dfq1 <- dfq[,c("fval", "value_new_q")]
  df_pq <- merge(dfp1, dfq1, by = "fval")
  #df_tq <- df_pq
  return(df_pq)
}

#' Q-Q plot comparing two samples with small discrete counts
#'
#' This function returns a ggplot object used to visualize quantiles comparing distributions of two samples.
#'
#' This is a function for quantile-quantile plot comparing comparing samples from two discrete distributions
#' after \emph{continuity correction} and linear interpolation
#'
#' @param P A numeric vector from one sample.
#' @param Q A numeric vector from the other sample.
#' @param sample1 A character to denote sample name of one distribution \code{P} generated from.
#' @param sample2 A character to denote sample name of the other distribution \code{Q} generated from.
#'
#' @return A ggplot object. Q-Q plot with continuity correction. Quantiles from one sample on the horizontal axis and corresponding quantiles
#' from the other sample on the vertical axis.
#'

qqplot_small_test <- function(P, Q, sample1, sample2){

  dfp <- new_quantile(P, sample1)
  dfq <- new_quantile(Q, sample2)

  df_tq <- qq_interpolation(dfp, dfq, sample1, sample2)
  ggplot(data = df_tq) +
    geom_line(data = df_tq, aes(x=.data$value_new_q, y=.data$value_new_p)) +
    geom_point(data = df_tq, aes(x=.data$value_new_q, y=.data$value_new_p)) +
    geom_abline(intercept = 0, slope = 1, color = "red") +
    xlab("Q") + ylab("P") +
    coord_fixed(ratio = 1)
}

#' A more "continuous" approximation of quantiles from the theoretical Poisson distribution.
#'
#' This function returns a data frame including the probability
#' from cumulative distribution function (CDF) and corresponding quantiles.
#'
#' This is a function developed to get corresponding quantiles from theoretical Poisson distribution.
#' The data points ranges from 0 to maximum value of sampled data used to compare with the theoretical Poisson distribution.
#'
#' @param data A numeric vector of sampled data points to compare with theoretical Poisson.
#' @param lambda A numeric value for theoretical Poisson distribution parameter (equal to mean).
#'
#' @return A data frame contains CDF probability and corresponding quantiles from the theoretical Poisson distribution.
#'
#'
new_quantile_pois <- function(data, lambda){

  if(!is.numeric(data)){
    warning("need numeric values for input data")
  }

  if(min(data < 0)){
    warning("Poisson random number should be non negative")
  }

  dfp <- new_quantile(data, "sample")
  dfq <- data.frame(value = seq(0:max(dfp$value)), fval = ppois(q=seq(0:max(dfp$value)), lambda = lambda))
  dfq <- dfq[with(dfq, order(fval)),]
  dfq$cat <- "theoretical"
  dfq <- rbind(data.frame(dfq), data.frame(value = c(-1, 0), cat = c("theoretical"), fval = c(0, dpois(0, lambda))))
  dfq$value_new <- dfq$value + 1/2
  dfq <- dfq[with(dfq, order(fval)), ]
  return(dfq)
}

#' Random sample generation function to generate sets of samples from theoretical Poisson distribution.
#'
#' This function returns a data frame with generated sets of samples and simulation index.
#'
#' This is a function used to simulate a given number sets of samples from a theoretical Poisson distribution
#' that match input samples on sample size and sample mean (or theoretical Poisson parameter).
#' Plotting these as envelopes in Q-Q plot shows the variability in shapes we can expect when
#' sampling from the theoretical Poisson distribution.
#'
#' @param x A numeric vector of sampled data points to compare with theoretical Poisson.
#' @param lambda A numeric value specifying mean for theoretical Poisson distribution.
#' @param R A numeric value specifying the number of simulated sets.
#'
#' @return A data frame contains simulated data and corresponding simulation index.

#' Random sample generation function to generate sets of samples from theoretical Poisson distribution.
#'
#' nboot_small returns a data frame with generate sets of samples and simulation index.
#'
#' This is a function used to simulate a given number sets of samples from a theoretical Poisson distribution
#' that match input samples on sample size and sample mean (or theoretical Poisson parameter).
#' Plotting these as envelopes in Q-Q plot shows the variability in shapes we can expect when
#' sampling from the theoretical Poisson distribution.
#'
#' @param x a numeric vector of sampled data points to compare with theoretical Poisson.
#' @param lambda a numeric value for mean of theoretical Poisson.
#' @param R a numeric value for mean of theoretical Poisson.
#'
#' @return a numeric vector of number of simulation sets that match input samples on sample size
#' and sample mean (or theoretical Poisson parameter).
#'
nboot_small <- function(x, lambda, R) {
  n <- length(x)
  do.call(rbind,
          lapply(1 : R,
                 function(i) {
                   xx <- sort(rpois(n, lambda))
                   data.frame(value = xx, sim = i)
                 }))
}

#' Q-Q plot comparing samples with a theoretical Poisson distribution
#'
#' This function returns a Q-Q plot with envelope using a more "continuous" approximation of quantiles.
#'
#' This is a function for Q-Q envelope plot used to compare whether given sample data points come from the
#' theoretical Poisson distribution.  By simulating repeated samples of the same size from the candidate
#' theoretical distribution, and overlaying the envelope on the same figure, it provides a feeling of
#' understanding the natural variation from the theoretical distribution.
#'
#' If an S3 object for class 'scppp' is used as input and the stored result under "data" is a matrix,
#' The GLM-PCA algorithm will be applied to estimate the Poisson parameter for each matrix entry.
#' Then a specific number of entries will be selected as sample data points to compare with the theoretical Poisson distribution.
#'
#' @param sample_data A numeric vector of sample data points or an S3 object for class 'scppp'.
#' @param lambda A numeric value specifying the theoretical Poisson parameter.
#' @param envelope_size A numeric value specifying the size of envelope on Q-Q plot (default 100).
#' @param ... not used.
#'
#' @return A ggplot object.
#'
#' @references
#' \insertRef{glmpca}{scpoisson}
#'
#' @import ggplot2
#'
#' @export
#'
qqplot_env_pois <- function(sample_data, lambda, envelope_size = 100, ...) {
  UseMethod("qqplot_env_pois")
}

#' @export
qqplot_env_pois.numeric <- function(sample_data, lambda, envelope_size = 100, ...){

  test_raw <- sample_data

  dfp <- new_quantile(test_raw, "sample")
  dfq <- new_quantile_pois(test_raw, lambda)

  df_tq <- qq_interpolation(dfp, dfq, "sample", "theoretical")
  df_tq <- subset(df_tq, df_tq$value_new_p >= (dfp$value[2]-0.5) & df_tq$value_new_p <= max(dfp$value)
                  & df_tq$value_new_q >= (dfp$value[2]-0.5) & df_tq$value_new_q <= max(dfp$value))
  df_tq$source <- "sample"

  test_all_qq <- df_tq

  gb <- nboot_small(sample_data, lambda, envelope_size)

  df_all <- purrr::map_dfr(
    .x = 1:envelope_size,
    .f = function(j, gb, lambda){
      dfp <- new_quantile(gb$value[which(gb$sim == j)], j)
      dfq <- new_quantile_pois(dfp$value, lambda)
      df_pq <- qq_interpolation(dfp, dfq, j, "theoretical")
      df_pq$sim <- j
      return(df_pq)
    }, gb = gb, lambda = lambda)
  df_all <- subset(df_all, df_all$value_new_p >= min(test_all_qq$value_new_p-1) & df_all$value_new_p <= (max(test_all_qq$value_new_q, test_all_qq$value_new_p) + 1)
                   & df_all$value_new_q >= min(test_all_qq$value_new_p-1) & df_all$value_new_q <= (max(test_all_qq$value_new_q, test_all_qq$value_new_p) + 1))

  min <- min(df_all$value_new_q, df_all$value_new_p, test_all_qq$value_new_q, test_all_qq$value_new_p)
  max <- max(df_all$value_new_q, df_all$value_new_p, test_all_qq$value_new_q, test_all_qq$value_new_p)
  #size <- 18
  p <- ggplot(data = df_all) +
    geom_line(aes(x = .data$value_new_q, y=.data$value_new_p, group = .data$sim),
              color = "gray", size = 1.2) +
    geom_line(data = test_all_qq, aes(x = .data$value_new_q, y = .data$value_new_p, color = source), size = 1.2) +
    geom_point(data = test_all_qq, aes(x = .data$value_new_q, y = .data$value_new_p, color = source), size = 1.5) +
    scale_color_manual(values=c("#E69F00")) +
    geom_abline(intercept = 0, slope = 1, color = "black", linetype = "dashed") +
    #coord_cartesian(xlim =c((dfp$value[2]-1), max(dfp$value + 1)), ylim = c((dfp$value[2]-1), max(dfp$value + 1))) +
    xlab("theoretical quantile") + ylab("sample quantile") + coord_fixed(ratio = 1) +
    labs(title = paste(expression("\u03BB"), "=", lambda)) +
    xlim(c(min, max)) + ylim(c(min, max))
  return(p)
}

#' @param L A numeric value specifying the number of latent vectors included when estimate the Poisson parameter for each matrix entry. This is not useful if a numeric vector is used as input.
#' @param select_by A character indicating whether entries should be selected by
#' \itemize{
#' \item{entry}: {independent matrix entries with Poisson parameter estimates closest to \code{lambda}}
#' \item{\code{cell}}: {one particular cell with UMI count mean closest to \code{lambda}}
#' \item{\code{gene}}: {one particular gene with UMI count mean closest to \code{lambda}}
#' }
#' This is not useful if a numeric vector is used as input.
#' @param entry_size A numeric value specifying the number of entries used to compare with the theoretical Poisson distribution.
#' This is not useful if a numeric vector is used as input, or the entries are selected by cell or gene.
#' @export
qqplot_env_pois.scppp <- function(sample_data, lambda, envelope_size = 100,
                                  L = 10,
                                  select_by = "entry",
                                  entry_size = 200, ...){
  test_data <- sample_data[["data"]]
  if(is.vector(test_data) && is.atomic(test_data)){
    qqplot_env_pois.numeric(test_data, lambda, envelope_size)
  }
  else {

    select_data <- function(test_dat, Gfit_dat, lambda,
                            select_by = "entry", entry_size = 200){

      if(select_by == "entry"){
        test_raw <- as.vector(test_dat)
        fitdata <- as.vector(Gfit_dat)
        order <- order(abs(fitdata - lambda), decreasing = F)
        test_raw <- test_raw[order]
        return(test_raw[1:entry_size])
      }

      if(select_by == "gene"){
        cell_mean <- rowMeans(test_dat)
        measure <- abs(cell_mean - lambda)
        index <- which(measure == min(measure))

        cell <- rownames(test_dat)[index]
        cell_sd <- matrixStats::rowSds(test_dat)[index]
        select_cell <- cell[which(cell_sd == max(cell_sd))]
        return(test_dat[which(rownames(test_dat) == select_cell[1]), ])
      }

      if(select_by == "cell"){
        gene_mean <- colMeans(test_dat)
        measure <- abs(gene_mean - lambda)
        index <- which(measure == min(measure))

        gene <- colnames(test_dat)[index]
        gene_sd <- matrixStats::colSds(test_dat)[index]
        select_gene <- gene[which(gene_sd == max(gene_sd))]
        return(test_dat[, which(colnames(test_dat) == select_gene[1])])
      }

    }

    apply_glmpca <- function(rawdata, L = 10){

      set.seed(1234)
      test_df <- rawdata
      test_df <- test_df[which(rowSums(test_df) > 0), ]
      ctl <- list(maxIter=500,eps=1e-4)
      res <- glmpca::glmpca(test_df, L=L, fam = "poi", sz=colSums(test_df),verbose=TRUE,ctl=ctl)
      factors <- res$factors

      U <- as.matrix(res$factors)
      V <- as.matrix(res$loadings)
      Ni <- colSums(test_df)
      Vj <- as.vector(as.matrix(res$coefX))
      Gfit_dat <- exp(t(t(U %*% t(V)) + Vj) + log(Ni))
      return(t(Gfit_dat))
    }

    test <- as.matrix(test_data)
    Gfit_dat <- apply_glmpca(test_data, L)
    sample_data <- select_data(test_data, Gfit_dat, lambda, select_by, entry_size)
    qqplot_env_pois.numeric(sample_data, lambda, envelope_size)
  }
}









